src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasInMemoryService.kt

```diff
@@ -29,6 +29,14 @@ class CosmasInMemoryService : CosmasGrpc.CosmasImplBase() {
 
     private val files = mutableMapOf<String, MutableList<ByteString>>()
 
+    private val versions = mutableMapOf<String, MutableList<storedStructure>>()

```

Comment:
What you are going to keep here is not versions. Versions are stored in the map above. Single edit is not a version.
I would call these things "patches" unless you have a better name. You have a base version, you apply patches.